File: Capitulo3- Funcoes/Funcoes.Funcional/Funcoes.Funcional/FuncaoAltaOrdem.fs
module FuncaoAltaOrdem

// Versão 1 do Código
//let imprimirNomes() =
//    let nomes = ["Gabriel"; "Rodrigo"; "Jhony"; "Luan"]
//    for nome in nomes do
//        printfn "Olá %s." nome

//let imprimirDobroDosNumeros numeros =
//    let numeros = [1..10]
//    for numero in numeros do
//        printfn "%i." ( numero * 2)


//Versão 2 do código
//let imprimirNomes nomes =
//    for nome in nomes do
//        printfn "Olá %s." nome
//
//let imprimirDobroDosNumeros numeros =
//    for numero in numeros do
//        printfn "%i." ( numero * 2)


let executarAcaoSobreElementos lista acao = lista |> List.iter acao
    
let imprimirNomes nomes = 
    executarAcaoSobreElementos nomes (printfn "Olá %s")    
    
let imprimirDobroDeUmNumero numero = 
    printfn "%i." ( numero * 2)

let multiplicarEimprimirNumero multiplicador numero = 
    printfn "%i." ( numero * multiplicador)

let imprimirDobroDosNumeros numeros = 
    executarAcaoSobreElementos numeros (imprimirDobroDeUmNumero)    

let imprimirDobroDosNumeros numeros = 
    executarAcaoSobreElementos numeros 
        (fun numero -> printfn "%i" (numero * 2) )    

let imprimirListaDeNumeros multiplicador numeros = 
    executarAcaoSobreElementos numeros 
        (fun numero -> 
             multiplicarEimprimirNumero multiplicador numero )    

File: Capitulo4- Funcoes - Parte 2/Funcoes.Funcional/Funcoes.NetFramework/FuncaoAltaOrdem.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Funcoes
{
    public static class FuncaoAltaOrdem
    {
        public static void ExecutarAcaoSobreElementos<TipoElemento>(List<TipoElemento> lista, Action<TipoElemento> acao)
        {
            lista.ForEach(acao);
        }

        public static void ImprimirNomes(List<string> nomes)
        {
            ExecutarAcaoSobreElementos(nomes, Console.WriteLine);
        }

        public static Action<int> MultiplicarEimprimirNumero(int multiplicador)
        {
            Action<int> imprimirNumero = 
                (numero) => Console.WriteLine(numero * multiplicador);
            return imprimirNumero;
        }

        public static void ImprimirNumeros(List<int> numeros, int multiplicador)
        {
            Action<int> multiplicaEDepoisEImprimi = 
                MultiplicarEimprimirNumero(multiplicador);

            ExecutarAcaoSobreElementos(numeros, multiplicaEDepoisEImprimi);
        }

        public static void Program()
        {
            Action<int> multiplicaPor5EDepoisImprimi = MultiplicarEimprimirNumero(5);
            Action<int> multiplicaPor3EDepoisImprimi = MultiplicarEimprimirNumero(3);
            int numeroQueSeraMultiplicado = 10;

            multiplicaPor5EDepoisImprimi(numeroQueSeraMultiplicado);
            multiplicaPor3EDepoisImprimi(numeroQueSeraMultiplicado);
        }
    }
}
